File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/VfsProviderClient.java
package app.zxtune.fs.provider;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import androidx.annotation.Nullable;

// TODO: use provider notifications instead of polling
public class VfsProviderClient {

  interface StatusCallback {
    void onProgress(int done, int total);
  }

  public interface ListingCallback extends StatusCallback {
    void onDir(Uri uri, String name, String description, @Nullable Integer icon, boolean hasFeed);

    void onFile(Uri uri, String name, String description, String details, @Nullable Integer tracks,
                @Nullable Boolean cached);
  }

  public interface ParentsCallback extends StatusCallback {
    void onObject(Uri uri, String name, @Nullable Integer icon);
  }

  private final ContentResolver resolver;

  public VfsProviderClient(Context ctx) {
    this.resolver = ctx.getContentResolver();
  }

  public final void resolve(Uri uri, ListingCallback cb) throws Exception {
    final Uri resolverUri = Query.resolveUriFor(uri);
    for (; ; ) {
      final Cursor cursor = resolver.query(resolverUri, null, null, null, null);
      if (cursor == null) {
        break;
      }
      try {
        if (!cursor.moveToFirst()) {
          break;
        } else if (Schema.Status.isStatus(cursor)) {
          getStatus(cursor, cb);
          waitOrCancel(resolverUri);
        } else {
          getObjects(cursor, cb);
          break;
        }
      } finally {
        cursor.close();
      }
    }
  }

  public final void list(Uri uri, ListingCallback cb) throws Exception {
    final Uri resolverUri = Query.listingUriFor(uri);
    for (; ; ) {
      final Cursor cursor = resolver.query(resolverUri, null, null, null, null);
      if (cursor == null) {
        break;
      }
      try {
        if (!cursor.moveToFirst()) {
          break;
        } else if (Schema.Status.isStatus(cursor)) {
          getStatus(cursor, cb);
          waitOrCancel(resolverUri);
        } else {
          getObjects(cursor, cb);
          break;
        }
      } finally {
        cursor.close();
      }
    }
  }

  public final void parents(Uri uri, ParentsCallback cb) throws Exception {
    final Uri resolverUri = Query.parentsUriFor(uri);
    for (; ; ) {
      final Cursor cursor = resolver.query(resolverUri, null, null, null, null);
      if (cursor == null) {
        break;
      }
      try {
        if (!cursor.moveToFirst()) {
          break;
        } else if (Schema.Status.isStatus(cursor)) {
          getStatus(cursor, cb);
          waitOrCancel(resolverUri);
        } else {
          getObjects(cursor, cb);
          break;
        }
      } finally {
        cursor.close();
      }
    }
  }

  public final void search(Uri uri, String query, ListingCallback cb) throws Exception {
    final Uri resolverUri = Query.searchUriFor(uri, query);
    for (; ;) {
      checkForCancel(resolverUri);
      final Cursor cursor = resolver.query(resolverUri, null, null, null, null);
      if (cursor == null) {
        return;
      }
      try {
        if (cursor.moveToFirst()) {
          if (Schema.Status.isStatus(cursor)) {
            getStatus(cursor, cb);
          } else {
            getObjects(cursor, cb);
            if (cursor.isLast()) {
              break;
            }
          }
        } else {
          waitOrCancel(resolverUri);
        }
      } finally {
        cursor.close();
      }
    }
  }

  public static Uri getFileUriFor(Uri uri) {
    return Query.fileUriFor(uri);
  }

  private void checkForCancel(Uri resolverUri) throws Exception {
    if (Thread.interrupted()) {
      resolver.delete(resolverUri, null, null);
      throw new InterruptedException();
    }
  }

  private void waitOrCancel(Uri resolverUri) throws Exception {
    try {
      Thread.sleep(1000);
    } catch (InterruptedException e) {
      resolver.delete(resolverUri, null, null);
      throw e;
    }
  }

  private static void getStatus(Cursor cursor, StatusCallback cb) throws Exception {
    final String err = Schema.Status.getError(cursor);
    if (err != null) {
      throw new Exception(err);
    } else {
      cb.onProgress(Schema.Status.getDone(cursor), Schema.Status.getTotal(cursor));
    }
  }

  private static void getObjects(Cursor cursor, ListingCallback cb) {
    do {
      if (Schema.Listing.isLimiter(cursor)) {
        break;
      }
      final Uri uri = Schema.Listing.getUri(cursor);
      final String name = Schema.Listing.getName(cursor);
      final String description = Schema.Listing.getDescription(cursor);
      if (Schema.Listing.isDir(cursor)) {
        cb.onDir(uri, name, description, Schema.Listing.getIcon(cursor),
            Schema.Listing.hasFeed(cursor));
      } else {
        cb.onFile(uri, name, description, Schema.Listing.getDetails(cursor),
            Schema.Listing.getTracks(cursor), Schema.Listing.isCached(cursor));
      }
    } while (cursor.moveToNext());
  }

  private static void getObjects(Cursor cursor, ParentsCallback cb) {
    do {
      final Uri uri = Schema.Parents.getUri(cursor);
      final String name = Schema.Parents.getName(cursor);
      final Integer icon = Schema.Parents.getIcon(cursor);
      cb.onObject(uri, name, icon);
    } while (cursor.moveToNext());
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/VfsProviderClient.kt
package app.zxtune.fs.provider

import android.content.Context
import android.database.ContentObserver
import android.database.Cursor
import android.net.Uri
import android.os.CancellationSignal
import app.zxtune.Releaseable
import app.zxtune.use

/**
 * Threads scheme
 *
 *  User            Binder1           VfsProviderClient    Binder2
 *
 *  queryListing -> Provider.query -> Observer.onChange -> Provider.query
 *                    |                                     |
 *  Callback.result <-+               Callback.onProgress <-+
 *                                    (may cancel query)
 *
 */
class VfsProviderClient(ctx: Context) {
    interface ListingCallback {
        fun onProgress(status: Schema.Status.Progress)
        fun onDir(dir: Schema.Listing.Dir)
        fun onFile(file: Schema.Listing.File)
    }

    interface ParentsCallback {
        fun onObject(obj: Schema.Parents.Object)
    }

    private val resolver = ctx.contentResolver

    @Throws(Exception::class)
    fun resolve(uri: Uri, cb: ListingCallback, signal: CancellationSignal? = null) =
        queryListing(Query.resolveUriFor(uri), cb, signal)

    private fun queryListing(
        resolverUri: Uri, cb: ListingCallback, userSignal: CancellationSignal?
    ) {
        val signal = userSignal ?: CancellationSignal()
        val notification = subscribeForChanges(resolverUri) {
            query(resolverUri)?.use {
                runCatching { getListing(it, cb) }.onFailure {
                    signal.cancel()
                    // Do not throw anything!!!
                }
            }
        }
        notification.use {
            query(resolverUri, signal)?.use {
                getListing(it, cb)
            }
        }
    }

    @Throws(Exception::class)
    fun list(uri: Uri, cb: ListingCallback, signal: CancellationSignal? = null) =
        queryListing(Query.listingUriFor(uri), cb, signal)

    @Throws(Exception::class)
    fun parents(uri: Uri, cb: ParentsCallback) {
        val resolverUri = Query.parentsUriFor(uri)
        query(resolverUri)?.use {
            getParents(it, cb)
        }
    }

    @Throws(Exception::class)
    fun search(uri: Uri, query: String, cb: ListingCallback, signal: CancellationSignal? = null) =
        queryListing(Query.searchUriFor(uri, query), cb, signal)

    // Notifications are propagated down-up, but since they sent to the root notifications
    // url, subscribe to it
    fun subscribeForNotifications(
        uri: Uri, cb: (Schema.Notifications.Object?) -> Unit
    ) = subscribeForChanges(Query.notificationUriFor(Uri.EMPTY)) {
        cb(getNotification(uri))
    }.also {
        cb(getNotification(uri))
    }

    fun getNotification(uri: Uri) = query(Query.notificationUriFor(uri))?.use {
        getNotification(it)
    }

    private fun query(uri: Uri, signal: CancellationSignal? = null) =
        resolver.query(uri, null, null, null, null, signal)

    private fun subscribeForChanges(uri: Uri, cb: () -> Unit): Releaseable {
        val observer = object : ContentObserver(null) {
            override fun onChange(selfChange: Boolean) = cb()
        }
        resolver.registerContentObserver(uri, false, observer)
        return Releaseable { resolver.unregisterContentObserver(observer) }
    }

    companion object {
        @JvmStatic
        fun getFileUriFor(uri: Uri, size: Long) = Query.fileUriFor(uri, size)

        private fun getListing(cursor: Cursor, cb: ListingCallback) {
            while (cursor.moveToNext()) {
                when (val obj = Schema.Object.parse(cursor)) {
                    is Schema.Listing.Dir -> cb.onDir(obj)
                    is Schema.Listing.File -> cb.onFile(obj)
                    is Schema.Status.Error -> throw Exception(obj.error)
                    is Schema.Status.Progress -> cb.onProgress(obj)
                    else -> Unit
                }
            }
        }

        private fun getParents(cursor: Cursor, cb: ParentsCallback): Boolean {
            while (cursor.moveToNext()) {
                when (val obj = Schema.Parents.Object.parse(cursor)) {
                    is Schema.Status.Error -> throw Exception(obj.error)
                    is Schema.Parents.Object -> cb.onObject(obj)
                    else -> return false
                }
            }
            return true
        }

        private fun getNotification(cursor: Cursor) = if (cursor.moveToNext()) {
            when (val obj = Schema.Notifications.Object.parse(cursor)) {
                is Schema.Status.Error -> throw Exception(obj.error)
                is Schema.Notifications.Object -> obj
                else -> null
            }
        } else {
            null
        }
    }
}
